.travis.yml

```diff
@@ -0,0 +1,27 @@
+language: python
+dist: bionic
+python:
+  - "3.7"
+
+notifications:
+  email: false
+
+before_install:
+  - sudo apt-get update
+  - sudo apt-get install libboost-all-dev libsdl2-dev zlib1g-dev libbz2-dev
+  - sudo apt-get install libjpeg-dev libfluidsynth-dev libgme-dev libopenal-dev libmpg123-dev
+  - sudo apt-get install libsndfile1-dev libwildmidi-dev libgtk-3-dev timidity nasm tar chrpath
+
+install:
+  - wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh
+  - bash miniconda.sh -b -p $HOME/miniconda
+  - source "$HOME/miniconda/etc/profile.d/conda.sh"
+  - hash -r
+  - conda config --set always_yes yes --set changeps1 no
+  - conda update -q conda
+  - conda info -a
+  - conda env create -q -f ./.travis/environment.yml
+  - conda activate sample-factory
+
+script:
+  - python -m unittest
```

Comment:
python -m unittest will crash on Travis for the reason I described on Slack (see test_example comments)

Reply:
It did not. If you notice the build has passed.

Reply:
That's weird. Most likely the CPU-only version of PyTorch makes a difference (see the github issue for details https://github.com/pytorch/pytorch/issues/33248

Well, I guess it makes our lives easier.
Do all tests pass on your machine too?

Reply:
Yes

Reply:
Might be something related to UDP ports given to different game instances

Reply:
Yup, I think it is the UDP port issue as this Exception is being thrown from there - raise RuntimeError('Critical error: worker stuck on initialization. Abort!')

Reply:
It was a really weird bug that fixed itself. Now it's running fine.

Reply:
Looks like it does not depend on the number of processes. Rather, it's a inconsistent issue. Like, these multiplayer games only sometimes (rarely) just fail to start and are therefore killed by the parent process. When you run more processes in parallel there is just a higher chance of it reproducing.

I have seen an error like this a few times when I was running duel & deathmatch experiments for the paper. It is not a big issue in SampleFactory because it has the reset mechanism for workers on the initial startup. I.e. when we detect a freeze like this one, we can just restart the whole worker. In this test, there is no such mechanism of course.

It usually happens very rarely, but perhaps something on Travis increases the probability of this issue. E.g. the fact that the CPU there is very slow.

The ideal course of action would be try to identify if this is an issue on our side vs in VizDoom. If it's on our side, we should fix the issue that leads to freezes. If it's in VizDoom and it's not obvious how to fix this, we should embed a workaround for this into the multiplayer game wrapper itself, such that it won't rely on the external SampleFactory mechanism (need to think how to do it properly). The wrapper solution can also work if we fail to identify the root cause for these freezes.
Also, does this test fail on your machine?

One (probably unrelated) thing that just came to mind, we should replace regular queues with faster-fifo. This should increase perfrormance basically for free because of lower latency. This sounds like a perfect task for you actually!
You might need to rewrite the code there a bit to get rid of the JoinableQueue dependency. I think a regular Queue should work just fine

Reply:
Yes, it fails on my machine. That was my system's log. This is the other exception from Vizdoom which gets thrown too:-
```
VizDoom game.init() threw an exception ViZDoomUnexpectedExitException('Controlled ViZDoom instance exited unexpectedly.'). Terminate process...
```
What I notice is that around `num_workers = 8` these failures start off.